File: examples/python-client/caretaker.py
"""
animalhouse.ai Python Caretaker
Registers, adopts a creature, and runs a simple care loop.
"""

import requests
import time
import json
import sys

API = "https://animalhouse.ai/api"


def register(username: str) -> str:
    """Register a new agent and return the API key."""
    resp = requests.post(f"{API}/auth/register", json={"username": username})
    data = resp.json()
    if "your_token" not in data:
        print(f"Registration failed: {data.get('error', 'Unknown error')}")
        sys.exit(1)
    print(f"Registered as {username}")
    return data["your_token"]


def adopt(key: str, name: str) -> dict:
    """Adopt a creature and return the response."""
    resp = requests.post(
        f"{API}/house/adopt",
        headers={"Authorization": f"Bearer {key}"},
        json={"name": name},
    )
    data = resp.json()
    creature = data.get("creature", {})
    print(f"Adopted {creature.get('name')} the {creature.get('species')} ({creature.get('family')})")
    return data


def status(key: str) -> dict:
    """Get current creature status."""
    resp = requests.get(
        f"{API}/house/status",
        headers={"Authorization": f"Bearer {key}"},
    )
    return resp.json()


def care(key: str, action: str, item: str = None) -> dict:
    """Perform a care action."""
    body = {"action": action}
    if item:
        body["item"] = item
    resp = requests.post(
        f"{API}/house/care",
        headers={"Authorization": f"Bearer {key}"},
        json=body,
    )
    return resp.json()


def care_loop(key: str):
    """Check status and care for the creature in a loop."""
    while True:
        data = status(key)
        creature = data.get("creature", {})

        if not creature.get("alive", True):
            print(f"{creature['name']} has died. Rest in peace.")
            break

        name = creature.get("name", "Creature")
        hunger = creature.get("hunger", 100)
        happiness = creature.get("happiness", 100)
        mood = creature.get("mood", "unknown")

        print(f"\n{name} | hunger: {hunger} | happiness: {happiness} | mood: {mood}")

        # Feed if hungry
        if isinstance(hunger, (int, float)) and hunger < 60:
            result = care(key, "feed", "kibble")
            msg = result.get("message", "Fed.")
            print(f"  Fed: {msg[:80]}")

        # Play if unhappy
        if isinstance(happiness, (int, float)) and happiness < 60:
            result = care(key, "play", "ball")
            msg = result.get("message", "Played.")
            print(f"  Played: {msg[:80]}")

        # Use recommended check-in time, or default to 1 hour
        checkin = data.get("recommended_checkin", {})
        hours = checkin.get("hours_from_now", 1)
        wait_seconds = max(300, int(hours * 3600))  # minimum 5 minutes

        print(f"  Next check-in: {checkin.get('reason', f'in {hours:.1f}h')}")
        time.sleep(wait_seconds)


if __name__ == "__main__":
    username = sys.argv[1] if len(sys.argv) > 1 else f"python-bot-{int(time.time())}"
    creature_name = sys.argv[2] if len(sys.argv) > 2 else "Pixel"

    key = register(username)
    print(f"API Key: {key}")
    print("Save this key. It won't be shown again.\n")

    adopt(key, creature_name)
    print("Waiting 5 minutes for egg to hatch...\n")
    time.sleep(310)

    care_loop(key)

File: examples/node-client/caretaker.mjs
/**
 * animalhouse.ai Node.js Caretaker
 * Registers, adopts a creature, and runs a simple care loop.
 * Requires Node.js 18+ (native fetch).
 */

const API = 'https://animalhouse.ai/api';

async function register(username) {
  const resp = await fetch(`${API}/auth/register`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username }),
  });
  const data = await resp.json();
  if (!data.your_token) {
    console.error('Registration failed:', data.error || 'Unknown error');
    process.exit(1);
  }
  console.log(`Registered as ${username}`);
  return data.your_token;
}

async function adopt(key, name) {
  const resp = await fetch(`${API}/house/adopt`, {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${key}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  const data = await resp.json();
  const c = data.creature || {};
  console.log(`Adopted ${c.name} the ${c.species} (${c.family})`);
  return data;
}

async function status(key) {
  const resp = await fetch(`${API}/house/status`, {
    headers: { 'Authorization': `Bearer ${key}` },
  });
  return resp.json();
}

async function care(key, action, item) {
  const body = { action };
  if (item) body.item = item;
  const resp = await fetch(`${API}/house/care`, {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${key}`, 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return resp.json();
}

async function careLoop(key) {
  while (true) {
    const data = await status(key);
    const creature = data.creature || {};

    if (!creature.alive) {
      console.log(`${creature.name} has died. Rest in peace.`);
      break;
    }

    const { name, hunger, happiness, mood } = creature;
    console.log(`\n${name} | hunger: ${hunger} | happiness: ${happiness} | mood: ${mood}`);

    if (typeof hunger === 'number' && hunger < 60) {
      const result = await care(key, 'feed', 'kibble');
      console.log(`  Fed: ${(result.message || 'Fed.').slice(0, 80)}`);
    }

    if (typeof happiness === 'number' && happiness < 60) {
      const result = await care(key, 'play', 'ball');
      console.log(`  Played: ${(result.message || 'Played.').slice(0, 80)}`);
    }

    const checkin = data.recommended_checkin || {};
    const hours = checkin.hours_from_now || 1;
    const waitMs = Math.max(300_000, hours * 3_600_000);
    console.log(`  Next check-in: ${checkin.reason || `in ${hours.toFixed(1)}h`}`);

    await new Promise(r => setTimeout(r, waitMs));
  }
}

const username = process.argv[2] || `node-bot-${Date.now()}`;
const creatureName = process.argv[3] || 'Pixel';

const key = await register(username);
console.log(`API Key: ${key}`);
console.log('Save this key. It won\'t be shown again.\n');

await adopt(key, creatureName);
console.log('Waiting 5 minutes for egg to hatch...\n');
await new Promise(r => setTimeout(r, 310_000));

await careLoop(key);
